### ISSUE TITLE
featureUtility provisioning installing features from maven repository 

### ISSUE THREAD

**Issue by herman-kailey** [OpenLiberty/docs#849]
We are suggesting to create a command reference guide for the featureUtility (https://github.com/OpenLiberty/open-liberty/issues/9227) command-line tool, similar to what has been documented for the server command (referenced here https://openliberty.io/docs/ref/command/). the documentation will consist of at least the following pages and information:

**Overview:**
Usage: featureUtility {installFeature|installServerFeatures|viewSettings|find|help} [options]

Actions:

    installFeature, if
	Install one feature or a list of space-separated features.

    installServerFeatures, isf
	Install the required features for a named server.

    find
	Find features in the repositories that are applicable to your configuration or view detailed information about features.

    help
	Print help information for the specified action.

    viewSettings
	View a template for configuring repositories or a proxy, or view and validate the settings for the configured repositories or proxy.

Options:
	Use help [actionName] for detailed option information of each action.

Additionally, there are environment variables that can be specified to modify the behavior of featureUtility:
FEATURE_REPO_URL: Overriding Maven Central with an on-prem Maven repository
FEATURE_REPO_USER: Username for FEATURE_REPO_URL user credentialing
FEATURE_REPO_PASSWORD: Password for FEATURE_REPO_URL user credentialing
FEATURE_LOCAL_REPO: Overriding local Maven repository
http_proxy: configuring outbound http proxy, e.g. http_proxy=http://server:port
https_proxy: configuring outbound https proxy, e.g. https_proxy=https://username:password@server:port

Environment variables can be specified in the shell environment or as properties in the properties file at ${wlp.install.dir}/etc/featureUtility.properties file. Variables specified in the featureUtility properties file will take precedence over variables of the same name in the shell environment.
below are the property names that can be specified in the featureUtility.properties file:

Define the Maven on-prem repository to override Maven Central, add a property as mavenCentralMirror.url=url, and set the properties mavenCentralMirror.user=userId and mavenCentralMirror.password=password if the repository requires a user name and password.

mavenCentralMirror.url will override the environment variable FEATURE_REPO_URL.
mavenCentralMirror.user will override the environment variable FEATURE_REPO_USER. 
mavenCentralMirror.password will override the environment variable FEATURE_REPO_ PASSWORD.

To define additional remote repositories, add properties as customRepoName.url=url
Each repository name must be unique.
The defined repositories are accessed in the order that they are specified in the featureUtility.properties file.
If a repository requires a user name and password, set the properties customRepoName.user=userId and customRepoName.password=password.
For example,
remoteRepo1.url=http://my-remote-server1/maven2
remoteRepo2.url=https://my-remote-server2/secure/maven2
remoteRepo2.user=operator
remoteRepo2.password={aes}KM8dhwcv892Ss1sawu9R+

proxyHost, proxyPort, proxyUser, and proxyPassword will override the environment variable http_proxy and https_proxy.
For example,
proxyHost=my.proxy.server.ibm.com
proxyPort=9080
proxyUser=admin
proxyPassword={aes}AH5NLyd7DfGb12pK17Pw+

featureLocalRepo will override the environment variable FEATURE_LOCAL_REPO.
For example,
featureLocalRepo=/usr/IBM/maven/repository

**InstallFeature:**
Usage:
	featureUtility installFeature [options] name...
	featureUtility if [options] name...

Description:
	Install one feature or a list of space-separated features.

Options:
    --noCache
	Features installed from a remote repository                          
	(for example, Maven Central) are not cached locally.                 

    --verbose
	Display additional debug information while the action runs.

    --acceptLicense
	Accepts the license agreement.

    name ...
	One or more features that are separated by a space. For features,    
	specify the feature shortname or the Maven coordinate. If you        
	specify a Maven coordinate, it must contain at least                 
	the group ID and the artifact ID                                     
	(for example, io.openliberty.features:mpHealth-2.0).                 
	The version defaults to the current runtime version and the          
	packaging type defaults to ESA if the version and packaging          
	are not specified.

**installServerFeature:**
Usage:
	featureUtility installServerFeatures [options] name...
	featureUtility isf [options] name...

Description:
	Install the required features for a named server.

Options:
    --noCache
	Features installed from a remote repository                          
	(for example, Maven Central) are not cached locally.

    --verbose
	Display additional debug information while the action runs.

    --acceptLicense
	Accepts the license agreement.

    name ...
	The name of an existing server in the same Liberty environment.

**Help:**
Usage:
	featureUtility help name
Description:
	Print help information for the specified action.
Options:
	name ...
	Specifies a featureUtility command action to show only the help for the action.

**find**
Usage:
	featureUtility find
Description:
	Find features in the repositories that are applicable to your configuration or view detailed information about features.

Options:
	name ...
	One or more features that are separated by a space. For features,    
	specify the feature shortname or the Maven coordinate. If you        
	specify a Maven coordinate, it must contain at least                 
	the group ID and the artifact ID                                     
	(for example, io.openliberty.features:mpHealth-2.0).                 
	The version defaults to the current runtime version and the          
	packaging type defaults to ESA if the version and packaging          
	are not specified.

**viewSettings:**
Usage:
	featureUtility viewSettings
Description:
	View the repositories and proxy settings.

Options:
	--viewvalidationmessages
	Use this option to display the detailed messages from the            
	validation of the configured featureUtility.properties file.           
	Each message contains an error code, the line number where the error 
	was found, and the cause of the error.

**Comment by chirp1** [comment]
To the writer that creates the documentation for the featureUtility command: Ensure that you follow the same structure for the topic as other commands, such as the securityUtility command. Also, determine whether the environment variables belong in the same topic as the command, or whether they belong in another topic.

**Comment by Charlotte-Holt** [comment]
@alvinso @herman-kailey Here's the first draft of this topic: https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility.html. Can you please review and add feedback to this issue? Thx!

**Comment by alvinso** [comment]
@Charlotte-Holt 

Thanks for the document.  I have one comment.  Thanks.

1) For the `find` action, can we change the option _name_ to _searchString_ and update the description?  The description of _searchString_ will be something like "Finds features that their name or description matches the _searchString_ from Maven Central and configured Maven repositories. If you do not specify a search string, the command searches for all available features."

**Comment by Charlotte-Holt** [comment]
@alvinso I incorporated your comments. Please let me know if any other updates are necessary. Otherwise, please leave a comment with your review/approval. Thx!

**Comment by alvinso** [comment]
@Charlotte-Holt It looks good, thanks.

**Comment by lauracowen** [comment]
I just realised that the [securityUtility command topic](https://draft-openlibertyio.mybluemix.net/docs/ref/general/#securityUtility.html) does the same as this one in terms of explicitly talking about actions and options. Can you see what you think of the server command format regarding this one too? I think we need to be consistent across all our command topics. The bit I don't like is explicitly talking about "actions". I know that's technically what they are and maybe I'm wrong but it feels like that's exposing the functionality of the command rather than focusing on what it provides.

So `server encode` provides something meaningfully quite different from `server createSSLCertificate`. Like the `server create` command provides something quite different from `server package` or `server debug`. Rather than asking the user to piece together the command + action + option + optionValue (server + package + --include + minify), I tried to group them more meaningfully. So command + option (`server package` + `--include=minify`) and describe them like that.

In the securityUtility and featureUtility topics, you've got a halfway position between the KC (which has separate topics for lists of actions and options) and the server command topics (which break out the server actions into separate topics but including their options). I think we need to bring them together in format.

I'm not sure what the best approach is, so happy to discuss it. My main dislike is having to define what "actions" are - I think that needs to be obvious and not need defining, or just part of the concept of the command.

**Comment by lauracowen** [comment]
I know you're still working on this but I just checked from perspective of reviewing the securityUtility topics.
- we have `featureName` in the options table here (like `actionName` in the security one). I think they're a bit different as featureName is more explanatory and the point of what you're doing (rather than being self-referential about the command like actionName is). I wonder if `featureName` should be shown in italics though to show it's not a literal value the way that, say, `--acceptLicense` is. Or maybe it doesn't matter as it doesn't have a `--` at the start?
- I know you probably haven't addressed all my comments above yet but one that I think matters is about removing the Admin Center usage example completely from this topic. It's not a good example and, in fact, Alasdair spotted that it was explicitly not meant to work (a bug in the code means it does weird things, converting your OL installation to WL, instead of just not working).
- In the Options table, the `--acceptLicense` option needs to be more explanatory because we don't have licences to accept in Open Liberty. The reason this is here is because it also appears in WebSphere Liberty and because if you install a feature in Open Liberty that has a different licence from Liberty, you get prompted. So Alasdair suggested something like  "Accepts the license agreement for the feature, if the feature has a different license from the Liberty the feature is installed into" (that last bit needs re-wording slightly for style but that's the basic idea.

- featureUtility installServerFeatures
  - Don't need all those usage examples - most of that is already described in the table below. Can probably just have the first one.

**Comment by Charlotte-Holt** [comment]
@lauracowen Thx for the review! I got in all your feedback. Main points addressed from above: 

- [X] we have featureName in the options table here (like actionName in the security one). I think they're a bit different as featureName is more explanatory and the point of what you're doing (rather than being self-referential about the command like actionName is). I wonder if featureName should be shown in italics though to show it's not a literal value the way that, say, --acceptLicense is. 
- [X] I know you probably haven't addressed all my comments above yet but one that I think matters is about removing the Admin Center usage example completely from this topic. 
- [X] In the Options table, the --acceptLicense option needs to be more explanatory because we don't have licences to accept in Open Liberty. The reason this is here is because it also appears in WebSphere Liberty and because if you install a feature in Open Liberty that has a different licence from Liberty, you get prompted. So Alasdair suggested something like "Accepts the license agreement for the feature, if the feature has a different license from the Liberty the feature is installed into" (that last bit needs re-wording slightly for style but that's the basic idea.
- [X] Don't need all those usage examples - most of that is already described in the table below. Can probably just have the first one.
- [X] The structure of the "Actions and options" section follows how commands were doc'd in the KC rather than how we doc'd the server commands in OL docs. The OL docs, rather than documenting how the functionality is structured in .bat/.sh files, try to focus on the meaning of the commands that the user experience. Rather than talking about actions and options, they're just commands that do some particular task and options you can apply to them.
- [X] The "Repository and proxy modifications" section feels like a task topic about how to set up the featureUtility command to install from a Maven repository of your own? I'm not against it being here but it feels a bit odd. 
- [X] name isn't an option; it's required (as in https://www.openliberty.io/docs/ref/command/#server-debug.html) so it shouldn't be in the table.
- [X] Also the context maybe needs explaining a bit more in the intro and where relevant in the descriptions and viewSettings section. You can use featureUtility to install features into Open Liberty from Maven Central (which is where we publish features to at every 4 weekly release). You can also choose to move all the features to a Maven repository that you host inside your company firewall (for example) and use featureUtility to install features from there instead.
- [X] Usage examples - there are too many to see and we don't need to show everything; just the most likely commands people are wanting to run. So can you remove: featureUtility installFeature io.openliberty.features:mpHealth-2.2 - it's not necessary to show both and the coordinates format is shown in the table already. We definitely don't need two --noCache examples. I'm not even sure we need one as it's fairly clear what that does from the table explanation. Remove the usage example about installing Admin Center as it's an IBM feature (commercial liberty), not an Open Liberty feature. If someone runs that command, their OL installation is automatically converted to a WebSphere (commercial) Liberty installation without warning.
- [X] featureUtility viewSettings isn't really an example. If you break this out into separate 'commands' instead of a section per action, this would just be a command with its own description.

Could you review again:
- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-find.html
- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-help.html
- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-installFeature.html
- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-installServerFeatures.html
- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-viewSettings.html


Question I had on this - I agreed that the **Repository and proxy modifications** section was awkward in these topics, so I broke it out into it's own topic: https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-modifications.html. Would it make sense to just nest this with the `featureUtility` commands in the navigation?

**Comment by lauracowen** [comment]
I like the repo info as a separate topic. Yes, I think I agree it makes most sense to keep it with the commands - I think it'd seem weird if it were in the main topics.

viewSettings topic - minor thing but where you say "view repository and proxy settings" it might be worth saying "view Maven repository and proxy settings" or even also add on "..when Maven repositories other than, or in addition to, Maven Central have been configured." or something like that.

See what you think of the comment above but I'll sign this off anyway.

**Comment by Charlotte-Holt** [comment]
@lauracowen I agree and made that update. Thx!

@Rwalls1 Can you please peer review the following topics that are associated with this issue:

- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-find.html
- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-help.html
- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-installFeature.html
- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-installServerFeatures.html
- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-viewSettings.html
- https://draft-openlibertyio.mybluemix.net/docs/ref/general/#featureUtility-modifications.html

**Comment by Rwalls1** [comment]
## Peer review
The topics look good, I just have a couple comments: 
### featureUtility installFeature
- Typo: Change "The featureUtility installFeature command install features. " to "The featureUtility installFeature command installs features. "
### featureUtility installServerFeatures
- I think you could change "Using this command prevents you from having to individually list out all the features that you want to install." to "Using this command prevents the need to individually list all the features that you want to install." This creates a more active voice and makes the sentence more concise.

**Comment by Charlotte-Holt** [comment]
@chirp1 Topics at 
- https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/command/featureUtility-find.html
- https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/command/featureUtility-help.html
- https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/command/featureUtility-installFeature.html
- https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/command/featureUtility-installServerFeatures.html
- https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/command/featureUtility-viewSettings.html
- https://draft-openlibertyio.mybluemix.net/docs/21.0.0.1/reference/command/featureUtility-modifications.html

**Comment by Charlotte-Holt** [comment]
Editing updates made:

- "The `featureUtility find` command finds features in the repositories that are applicable to your configuration and allows you to view detailed information about them." --> "The `featureUtility find` command finds features in the repositories that are applicable to your configuration and allows you to view detailed information about these features." - DQTI, Chapter 6, Clarity, Use pronouns correctly
- "You can easily see Maven repository and proxy settings that are contained in the `featureUtility.properties` file by running the [`featureUtility viewSettings` command](https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/reference/command/featureUtility-viewSettings.html)." --> "You can easily see Maven repository and proxy settings that are contained in the `featureUtility.properties` file by running the [featureUtility viewSettings command](https://draft-openlibertyio.mybluemix.net/docs/20.0.0.11/reference/command/featureUtility-viewSettings.html)." - DQTI, Chapter 8, Style, Use consistent markup tagging

**Comment by chirp1** [comment]
Hi Charlotte,
- Global comment for the files. Where you have "path_to_liberty/bin", ensure that you add italics for "path_to_liberty" since it is a variable. Also, other variables, such as those in the syntax should be italic.
- featureUtility-find.html: For "searchString", check server commands etc as to how to word discussions about variables. Somewhere in there you want to say  the English wording of "search string". You might be able to do saying the variable itself, but if you do say it, put a noun after it. **chirp1: Looks OK in PR**
- featureUtility-help.html is OK.
- featureUtility-installFeature.html: Combine the first two paragraphs into one paragraph and have a total of just two sentences. **chirp1: Looks OK in PR** The rest of the topic looks OK.
- featureUtility-installServerFeatures.html: Reword the first two paragraphs so that in the first paragraph you include the point that you can install multiple features when you run the command once. The rest of the topic looks OK. **chirp1: How about rewording PR to something like "The featureUtility installServerFeatures command installs the required features that are listed in the sever.xml file of a server.?**  **chirp1: Looks OK in PR**
- featureUtility-viewSettings.html: (1) When you say "parsed properties" in the intro, is that the same as the "settings" mentioned in the info? If so, use terms consistently. **chirp1: Looks OK in PR** (2)For the usage example, I think that the settings would still be displayed in additions to the error messages. Would you clarify with the text one way or the other whether the settings would display for the user. **chirp1: Looks OK in PR**
- featureUtility-modifications.html: (1) I think that for the FEATURE_LOCAL_REPO environment variable there is an inconsistency. I seem to see more than one corresponding property for it. **Chirp1: OK in PR**  (2) For "It can be helpful to ", "it" is vague (one of the things on the cheat sheet to avoid). Reword. You might try a wording with "You can...".**Chirp1: OK in PR**

**Comment by chirp1** [comment]
Based on the PR, the updates look good. Charlotte is opening an issue for my global comment about italicizing variables. Adding Editorial reviewed label.

**Comment by Charlotte-Holt** [comment]
Validated on staging. Merged to `vNext` in #3441.
